components/sync15/README.md

```diff
@@ -5,4 +5,108 @@ data stores that want to sync against a Firefox Sync v1.5 sync server.
 It handles things like encrypting/decrypting records, obtaining and
 using storage node auth tokens, and so-on.
 
-It also needs a more complete README...
+There are 2 key concepts to understand here - the implementation itself, and
+a rust trait for a "syncable store" where component-specific logic lives - but
+before we dive into them, some preamble might help put things into context.
+
+## Introduction and History
+
+For many years Sync has worked exclusively against a "sync v1.5 server". This
+[is a REST API described here](https://mozilla-services.readthedocs.io/en/latest/storage/apis-1.5.html).
+The important part is that the API is conceptually quite simple - there are
+arbitrary "collections", indexed by a GUID, and lacking traditonal database
+concepts like joins. Because the payload is encrypted, there's very little
+scope for it to be much smarter. Thus it's reasonably easy to create a fairly
+generic abstraction over the API that can be easily reused.
+
+Back in the deep past, we found ourselves with 2 different components that
+needed to sync against a sync v1.5 server. The apps using these components
+didn't have schedulers or any UI for choosing what to sync - so these
+components just looked at the existing state of the engines on the server and
+synced if they were enabled.
+
+This was also pre-megazord - the idea was that apps could choose from a "menu"
+of components to include - so we didn't really want to bind these components
+together. Therefore, there was no concept of "sync all" - instead, each of the
+components had to be synced individually. So this component stated out as more
+of a "library" than a "component" which individual components could reuse - and
+each of these components was a "syncable store".
+
+Fast forward to Fenix and we needed a UI for managing all the engines supported
+there, and a single "sync now" experience etc - so we also have a sync_manager
+component - [see its README for more](../components/sync_manager/README.md).
+But even though it exists, there are still some parts of this component that
+reflect these early days - for example, it's still possible to sync just a
+single component using sync15 (ie, without going via the "sync manager"),
+although this isn't used and should be removed - the "sync manager" allows you
+to choose which engines to sync, so that should be used exclusively.
+
+## Metadata
+
+There's some metadata associated with a sync. Some of the metadata is "global"
+to the app (eg, the enabled state of engines, information about what servers to
+use, etc) and some is specific to an engine/store (eg, timestamp of the
+server's collection for this engine/store, guids for the collections, etc).
```

Comment:
Hrm, I'm not a fan of the way that "engine/store" keeps showing up here, since we explicitly talk about a trait named `Store` but we don't really say what an "engine" is. Is this assumed knowledge from elsewhere that we should link out to? Or maybe the distinction isn't that important at this level and we could just say "store" consistently?

Reply:
That's a really good point, so I ended up adding an entire section dedicated to this - so I'm going to re-request review from both of you, specifically for 0b3ba8075b56b79a811773c46d8338562fad55ee (which I'll squash before merging, but left it separate for review)